File: 2015/day05.cpp
// https://adventofcode.com/2015/day/5

#include "../common/common.h"

#include <algorithm>
#include <array>
#include <iostream>
#include <map>
#include <print>
#include <ranges>
#include <string>
#include <string_view>

constexpr bool contains_three_vowels(std::string_view str) {
  int count = 0;
  for (char c : str) {
    count += static_cast<int>(std::ranges::contains("aeiou", c));
    if (count >= 3) {
      return true;
    }
  }
  return false;
}

using namespace std::string_view_literals;
static constexpr std::array bad_strings = {"ab"sv, "cd"sv, "pq"sv, "xy"sv};

constexpr bool double_letters_ok(std::string_view str) {
  bool repeated = false;
  for (auto&& pair_raw : str | std::views::slide(2)) {
    const auto pair = std::string_view{pair_raw};
    if (std::ranges::contains(bad_strings, pair)) {
      return false;
    }
    repeated |= (pair[0] == pair[1]);
  }
  return repeated;
}

constexpr bool is_nice1(std::string_view str) {
  return contains_three_vowels(str) && double_letters_ok(str);
}

constexpr bool is_nice2(const std::string_view str) {
  bool pair_repeats = false;
  bool letter_repeats = false;
  // Skip last char
  for (const auto [index, c] :
       str.substr(0, str.size() - 1) | std::views::enumerate) {
    const auto pair = str.substr(index, 2);
    const auto rest = str.substr(index + 2);
    pair_repeats |= rest.contains(pair);
    letter_repeats |= (rest.size() > 0) && (c == rest[0]);
  }
  return pair_repeats && letter_repeats;
}

static_assert(is_nice2("qjhvhtzxzqqjkmpb"));
static_assert(is_nice2("xxyxx"));
static_assert(!is_nice2("uurcxstgmygtbstg"));
static_assert(!is_nice2("ieodomkazucvgmuy"));

int solve_case1(const std::string& filename) {
  return std::ranges::count_if(aoc::views::read_lines(filename), &is_nice1);
}

int solve_case2(const std::string& filename) {
  return std::ranges::count_if(aoc::views::read_lines(filename), &is_nice2);
}

int main() {
  std::println("Part 1");
  AOC_EXPECT_RESULT(2, solve_case1("day05.example"));
  AOC_EXPECT_RESULT(258, solve_case1("day05.input"));
  std::println("Part 2");
  AOC_EXPECT_RESULT(1, solve_case2("day05.example"));
  AOC_EXPECT_RESULT(53, solve_case2("day05.input"));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2015/src/day05.rs
use std::io;

fn contains_three_vowels(s: &str) -> bool {
    let mut count = 0;
    for c in s.chars() {
        count += "aeiuo".contains(c) as i32;
        if count >= 3 {
            return true;
        }
    }
    return false;
}

const BAD_STRINGS: [&str; 4] = ["ab", "cd", "pq", "xy"];

fn double_letters_ok(s: &str) -> bool {
    let mut repeated = false;
    for (first, second) in s.chars().zip(s.chars().skip(1)) {
        if BAD_STRINGS.contains(&String::from_iter([first, second]).as_str()) {
            return false;
        }
        repeated |= first == second;
    }
    return repeated;
}

fn is_nice1(s: &str) -> bool {
    contains_three_vowels(s) && double_letters_ok(s)
}

fn is_nice2(s: &str) -> bool {
    let mut pair_repeats = false;
    let mut letter_repeats = false;
    for (index, c) in s[0..s.len() - 1].chars().enumerate() {
        let pair = &s[index..index + 2];
        let rest = &s[index + 2..];
        pair_repeats |= rest.contains(pair);
        letter_repeats |= (rest.len() > 0) && (c == rest.chars().next().unwrap());
    }
    pair_repeats && letter_repeats
}

fn solve_case1(filename: &str) -> io::Result<i32> {
    Ok(std::fs::read_to_string(filename)?
        .lines()
        .filter(|s| is_nice1(s))
        .count() as i32)
}

fn solve_case2(filename: &str) -> io::Result<i32> {
    Ok(std::fs::read_to_string(filename)?
        .lines()
        .filter(|s| is_nice2(s))
        .count() as i32)
}

fn main() {
    println!("Part 1");
    assert_eq!(2, solve_case1("day05.example").unwrap());
    assert_eq!(258, solve_case1("day05.input").unwrap());
    println!("Part 2");
    assert_eq!(1, solve_case2("day05.example").unwrap());
    assert_eq!(53, solve_case2("day05.input").unwrap());
}
